File: design-patterns/csharp/adapter/CheckoutService.cs
using System;

namespace Adapter
{
    public class CheckoutService
    {
        private readonly IPaymentProcessor paymentProcessor;

        public CheckoutService(IPaymentProcessor paymentProcessor)
        {
            this.paymentProcessor = paymentProcessor;
        }

        public bool Checkout(double amount, string currency)
        {
            Console.WriteLine($"CheckoutService: Initiating checkout for {currency} {amount}");
            
            paymentProcessor.ProcessPayment(amount, currency);
            
            if (paymentProcessor.IsPaymentSuccessful())
            {
                Console.WriteLine($"CheckoutService: Payment successful. Transaction ID: {paymentProcessor.GetTransactionId()}");
                return true;
            }
            
            Console.WriteLine("CheckoutService: Payment failed.");
            return false;
        }
    }
} 

File: design-patterns/python/adapter/checkout_service.py
from .payment_processor import PaymentProcessor


class CheckoutService:
    """Client - Uses the PaymentProcessor interface"""
    
    def __init__(self, payment_processor: PaymentProcessor):
        self._payment_processor = payment_processor

    def checkout(self, amount: float, currency: str) -> None:
        print(f"CheckoutService: Attempting to process order for ${amount} {currency}")
        self._payment_processor.process_payment(amount, currency)
        if self._payment_processor.is_payment_successful():
            print(f"CheckoutService: Order successful! Transaction ID: {self._payment_processor.get_transaction_id()}")
        else:
            print("CheckoutService: Order failed. Payment was not successful.") 